src/flexmeasures_client/s2/control_types/FRBC/utils.py

```diff
@@ -0,0 +1,96 @@
+from datetime import datetime, timedelta
+from math import isclose
+from typing import List
+
+import pandas as pd
+from python_s2_protocol.FRBC.messages import FRBCInstruction, FRBCSystemDescription
+from python_s2_protocol.FRBC.schemas import FRBCOperationMode
+
+from flexmeasures_client.s2.utils import get_unique_id
+
+
+def compute_factor(operation_mode: FRBCOperationMode, fill_level, fill_rate):
+    """
+    Compute the operation mode factor for a fill_rate
+    """
+
+    for element in operation_mode.elements:
+        start_fill_level = element.fill_level_range.start_of_range
+        end_fill_level = element.fill_level_range.end_of_range
+
+        # potential optimization if we assume that the operation_mode_elements are
+        # sorted by fill_level
+        # if fill_level < start_fill_rate:
+        #   return  # TODO: raise error -> cannot find
+
+        # TODO: IMPLEMENT A METHOD `in` to check if a value is within a range
+        if (fill_level >= start_fill_level) and (fill_level <= end_fill_level):
+            start_fill_rate = element.fill_rate.start_of_range
+            end_fill_rate = element.fill_rate.end_of_range
+            delta_fill_rate = end_fill_rate - start_fill_rate
+
+            if (fill_rate >= start_fill_rate) and (fill_rate <= end_fill_rate):
+                operation_mode_factor = (fill_rate - start_fill_rate) / delta_fill_rate
+                return operation_mode_factor
+
+            else:
+                print("Requested fill rate out of range")
+                return
+
+    print("Couldn't find a valid fill_level")
+
+    return  # handle the case where we are given an invalid fill_level
+
+
+def fm_schedule_to_instructions(
+    schedule, system_description: FRBCSystemDescription, initial_fill_level: float
+) -> List[FRBCInstruction]:
+    values = schedule.get("values")
+
+    if len(values) == 0:
+        return []
+
+    print(schedule)
+
+    previous_value = None
+    instructions = []
+
+    start = datetime.fromisoformat(schedule.get("start"))
+    deltaT = pd.Timedelta(schedule.get("duration")) / len(values)
+
+    # assuming there's only 1 actuator
+    actuator = system_description.actuators[0]
+
+    # assuming there's only 1 operation mode in the actuator
+    operation_mode = actuator.operation_modes[0]
```

Comment:
Should we `raise NotImplementedError` otherwise?

Reply:
Instead, I'm returning an empty list of instructions.